index.rst

```diff
@@ -447,6 +447,64 @@ in some contexts where materializing the set would be expensive. The
 'DataRefSet' is usually generated by listing existing datasets of a particular
 dataset type, but its component 'DataRef's can be used with other dataset types.
 
+Caching
+-------
+
+When an object is read by the Butler, the Butler keeps a weakref to that object
+and returns a normal reference to that object. As long as the normal ref is kept
+(i.e. as long as the object is not garbage collected) the Butler will keep the
+weakref. If that object is read again, the butler will get a normal ref (via the
+weakref) from the cache instead of reading the object from persisted data a
+second time.
+
+This is accomplished by keeping a WeakValueDictionary in Butler, where the key
+is a custom hash of the ButlerLocation passed to Butler._read, and the value is
+the object that was loaded using that ButlerLocation. The hash of ButlerLocation
+parameters includes the items needed to uniquely identify an object to be loaded
+(or retrieved from the cache).
```

Comment:
I wonder if it's actually desirable to hash the whole `ButlerLocation` rather than just the `datasetType` and data ID. when we have multiple storages active in the same butler, the `datasetType` and data ID should be sufficient to uniquely identify a dataset, and which storage its loaded from should be an implementation detail.  Including the `pythonType` and `cppType` should be entirely redundant with the `datasetType`.  Is the problem that there may be other ways (aliases? composites?) to access some datasets with different `datasetType`s and/or data IDs?